Firestore/Source/API/FIRCollectionReference.mm

```diff
@@ -18,53 +18,34 @@
 
 #include <utility>
 
-#include "Firestore/core/src/firebase/firestore/util/autoid.h"
-
 #import "Firestore/Source/API/FIRDocumentReference+Internal.h"
 #import "Firestore/Source/API/FIRFirestore+Internal.h"
 #import "Firestore/Source/API/FIRQuery+Internal.h"
+#import "Firestore/Source/API/FSTUserDataConverter.h"
 #import "Firestore/Source/Core/FSTQuery.h"
 
+#include "Firestore/core/src/firebase/firestore/api/collection_reference.h"
 #include "Firestore/core/src/firebase/firestore/api/input_validation.h"
-#include "Firestore/core/src/firebase/firestore/model/document_key.h"
 #include "Firestore/core/src/firebase/firestore/model/resource_path.h"
-#include "Firestore/core/src/firebase/firestore/util/hashing.h"
+#include "Firestore/core/src/firebase/firestore/util/error_apple.h"
 #include "Firestore/core/src/firebase/firestore/util/string_apple.h"
 
 namespace util = firebase::firestore::util;
 using firebase::firestore::api::ThrowInvalidArgument;
-using firebase::firestore::model::DocumentKey;
 using firebase::firestore::model::ResourcePath;
-using firebase::firestore::util::CreateAutoId;
 
 NS_ASSUME_NONNULL_BEGIN
 
-@interface FIRCollectionReference ()
-- (instancetype)initWithPath:(const ResourcePath &)path
-                   firestore:(FIRFirestore *)firestore NS_DESIGNATED_INITIALIZER;
-
-// Mark the super class designated initializer unavailable.
-- (instancetype)initWithQuery:(api::Query &&)query NS_UNAVAILABLE;
-@end
-
-@implementation FIRCollectionReference (Internal)
-+ (instancetype)referenceWithPath:(const ResourcePath &)path firestore:(FIRFirestore *)firestore {
-  return [[FIRCollectionReference alloc] initWithPath:path firestore:firestore];
-}
-@end
-
 @implementation FIRCollectionReference
 
-- (instancetype)initWithPath:(const ResourcePath &)path firestore:(FIRFirestore *)firestore {
-  if (path.size() % 2 != 1) {
-    ThrowInvalidArgument("Invalid collection reference. Collection references must have an odd "
-                         "number of segments, but %s has %s",
-                         path.CanonicalString(), path.size());
-  }
+- (instancetype)initWithReference:(api::CollectionReference &&)reference {
+  return [super initWithQuery:std::move(reference)];
```

Comment:
This is somewhat controversial: this passes an rvalue reference of type `api::Query` with an instance of `api::CollectionReference`. This works because of the inheritance relationship and the fact that `api::CollectionReference` has no members (besides what's in its parent).